File: 005_rotate_and_translate_image/CPP/rotate_image/rotate_image.cpp
#include <iostream>
#include<opencv2/opencv.hpp>
using namespace cv;

int main(int, char**) 
{
    Mat image = imread("image.jpg");
	double angle = 45;

	// get the center coordinates of the image to create the 2D rotation matrix
	Point2f center((image.cols - 1) / 2.0, (image.rows - 1) / 2.0);
	// create the rotation matrix using the image center
	Mat rotation_matix = getRotationMatrix2D(center, angle, 1.0);

	// we will save the resulting image in rotated_image matrix
	Mat rotated_image;
	// apply affine transformation to the original image using the 2D rotaiton matrix
	warpAffine(image, rotated_image, rotation_matix, image.size());
	imshow("Original image", image);
	imshow("Rotated image", rotated_image);
	waitKey(0);
	// save the rotated image to disk
	imwrite("rotated_im.jpg", rotated_image);

	return 0;
}

File: 005_rotate_and_translate_image/Python/rotate_image.py
import cv2

# Reading the image
image = cv2.imread('image.jpg')

# dividing height and width by 2 to get the center of the image
height, width = image.shape[:2]
center = (width/2, height/2)

# the above center is the center of rotation axis
# using cv2.getRotationMatrix2D() to get the rotation matrix
rotate_matrix = cv2.getRotationMatrix2D(center=center, angle=45, scale=1)

# rotate the image using cv2.warpAffine
rotated_image = cv2.warpAffine(src=image, M=rotate_matrix, dsize=(width, height))

cv2.imshow('Original image', image)
cv2.imshow('Rotated image', rotated_image)
# wait indefinitely, press any key on keyboard to exit
cv2.waitKey(0)
# we can also write the rotated image to disk
cv2.imwrite('rotated_image.jpg', rotated_image)